File: collectors/instagram_follower_enricher.mjs
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";

const USER_FIELDS = [
  "user_id",
  "username",
  "first_seen_at",
  "last_seen_at",
  "follower_count",
  "follower_count_collected_at",
  "lookup_status",
  "last_lookup_at",
  "last_error",
];

const FOLLOWER_LOOKUP_FIELDS = [
  "collected_at",
  "user_id",
  "username",
  "follower_count",
  "error",
];

function csvValue(value) {
  const text = String(value ?? "").replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  return `"${text.replace(/"/g, '""')}"`;
}

function parseCsv(text) {
  const rows = [];
  let row = [];
  let value = "";
  let quoted = false;
  for (let index = 0; index < text.length; index += 1) {
    const character = text[index];
    if (quoted) {
      if (character === '"' && text[index + 1] === '"') {
        value += '"';
        index += 1;
      } else if (character === '"') {
        quoted = false;
      } else {
        value += character;
      }
    } else if (character === '"') {
      quoted = true;
    } else if (character === ",") {
      row.push(value);
      value = "";
    } else if (character === "\n") {
      row.push(value);
      rows.push(row);
      row = [];
      value = "";
    } else if (character !== "\r") {
      value += character;
    }
  }
  if (value || row.length) {
    row.push(value);
    rows.push(row);
  }
  return rows.filter((candidate) => candidate.some((cell) => cell !== ""));
}

function csvObjects(text) {
  const rows = parseCsv(String(text ?? "").replace(/^\uFEFF/, ""));
  if (!rows.length) return [];
  const headers = rows[0];
  return rows.slice(1).map((row) => Object.fromEntries(
    headers.map((header, index) => [header, row[index] ?? ""]),
  ));
}

async function atomicWriteCsv(destination, rows, fields) {
  await fsp.mkdir(path.dirname(destination), { recursive: true });
  const content = [
    `\uFEFF${fields.map(csvValue).join(",")}`,
    ...rows.map((row) => fields.map((field) => csvValue(row[field])).join(",")),
  ].join("\r\n") + "\r\n";
  const temporary = path.join(
    path.dirname(destination),
    `.${path.basename(destination)}.${process.pid}.${Date.now()}.tmp`,
  );
  await fsp.writeFile(temporary, content, "utf8");
  await fsp.rename(temporary, destination);
}

async function appendCsv(destination, row, fields) {
  await fsp.mkdir(path.dirname(destination), { recursive: true });
  const needsHeader = !fs.existsSync(destination) || (await fsp.stat(destination)).size === 0;
  const lines = [];
  if (needsHeader) lines.push(`\uFEFF${fields.map(csvValue).join(",")}`);
  lines.push(fields.map((field) => csvValue(row[field])).join(","));
  await fsp.appendFile(destination, `${lines.join("\r\n")}\r\n`, "utf8");
}

async function normalizeCsvSchema(destination, fields) {
  if (!fs.existsSync(destination)) return;
  const text = await fsp.readFile(destination, "utf8");
  const rows = parseCsv(text.replace(/^\uFEFF/, ""));
  const headers = rows[0] ?? [];
  if (headers.join("\u0000") === fields.join("\u0000")) return;
  const normalized = csvObjects(text).map((row) => Object.fromEntries(
    fields.map((field) => [field, row[field] ?? ""]),
  ));
  await atomicWriteCsv(destination, normalized, fields);
}

function parseInstagramJson(rawText) {
  const protectedIntegers = String(rawText ?? "").replace(
    /("(?:pk|pk_id|id)"\s*:\s*)(\d{16,})(?=\s*[,}])/g,
    '$1"$2"',
  );
  return JSON.parse(protectedIntegers);
}

class FollowerEnricher {
  constructor({
    dataDir,
    concurrency = 1,
    cacheHours = 1,
    lookupImpl,
    source = "instagram_web",
    onProgress = null,
    now = () => new Date(),
  }) {
    this.dataDir = path.resolve(dataDir);
    this.usersPath = path.join(this.dataDir, "users.csv");
    this.lookupsPath = path.join(this.dataDir, "follower_lookups.csv");
    if (typeof lookupImpl !== "function") {
      throw new TypeError("FollowerEnricher requires a web lookup function.");
    }
    this.concurrency = Math.max(1, Math.min(10, Math.trunc(Number(concurrency) || 1)));
    this.cacheHours = Math.max(0, Number(cacheHours) || 0);
    this.lookupImpl = lookupImpl;
    this.source = source;
    this.onProgress = onProgress;
    this.now = now;
    this.users = new Map();
    this.userIdIndex = new Map();
    this.usernameIndex = new Map();
    this.queue = [];
    this.queued = new Set();
    this.active = 0;
    this.stopped = false;
    this.writeChain = Promise.resolve();
    this.lookupWriteChain = Promise.resolve();
    this.usersDirtyChanges = 0;
    this.usersFlushTimer = null;
    this.drainWaiters = [];
    this.stats = {
      queued: 0,
      success: 0,
      unavailable: 0,
      failed: 0,
      completed: 0,
      stopStatus: "",
      stopError: "",
    };
    this.ready = this._load();
  }

  async _load() {
    await fsp.mkdir(this.dataDir, { recursive: true });
    await normalizeCsvSchema(this.usersPath, USER_FIELDS);
    await normalizeCsvSchema(this.lookupsPath, FOLLOWER_LOOKUP_FIELDS);
    if (!fs.existsSync(this.usersPath)) return;
    const rows = csvObjects(await fsp.readFile(this.usersPath, "utf8"));
    for (const source of rows) {
      const row = Object.fromEntries(USER_FIELDS.map((field) => [field, source[field] ?? ""]));
      const key = row.user_id ? `id:${row.user_id}` : `username:${row.username.toLowerCase()}`;
      if (key.endsWith(":")) continue;
      row._key = key;
      this.users.set(key, row);
      this._indexRow(row);
    }
  }

  _indexRow(row, previousUsername = "") {
    if (row.user_id) this.userIdIndex.set(row.user_id, row);
    if (previousUsername && this.usernameIndex.get(previousUsername.toLowerCase()) === row) {
      this.usernameIndex.delete(previousUsername.toLowerCase());
    }
    if (row.username) this.usernameIndex.set(row.username.toLowerCase(), row);
  }

  _findOrCreate(userId, username, seenAt) {
    let row = (userId && this.userIdIndex.get(userId))
      || (username && this.usernameIndex.get(username.toLowerCase()));
    if (!row) {
      const key = userId ? `id:${userId}` : `username:${username.toLowerCase()}`;
      row = Object.fromEntries(USER_FIELDS.map((field) => [field, ""]));
      row._key = key;
      row.first_seen_at = seenAt;
      this.users.set(key, row);
    }
    const previousUsername = row.username;
    if (userId) row.user_id = userId;
    if (username) row.username = username;
    row.first_seen_at ||= seenAt;
    row.last_seen_at = seenAt;
    this._indexRow(row, previousUsername);
    return row;
  }

  _flushUsers() {
    if (!this.usersDirtyChanges) return this.writeChain;
    this.usersDirtyChanges = 0;
    if (this.usersFlushTimer) {
      clearTimeout(this.usersFlushTimer);
      this.usersFlushTimer = null;
    }
    this.writeChain = this.writeChain.then(async () => {
      const rows = [...this.users.values()]
        .sort((left, right) => (left.first_seen_at || "").localeCompare(right.first_seen_at || ""))
        .map((row) => Object.fromEntries(USER_FIELDS.map((field) => [field, row[field] ?? ""])));
      await atomicWriteCsv(this.usersPath, rows, USER_FIELDS);
    });
    return this.writeChain;
  }

  _markUsersDirty() {
    this.usersDirtyChanges += 1;
    if (this.usersDirtyChanges >= 100) {
      void this._flushUsers().catch(() => {});
    } else if (!this.usersFlushTimer) {
      this.usersFlushTimer = setTimeout(() => {
        this.usersFlushTimer = null;
        void this._flushUsers().catch(() => {});
      }, 60_000);
    }
  }

  _scheduleLookupWrite(row) {
    this.lookupWriteChain = this.lookupWriteChain.then(() => appendCsv(
      this.lookupsPath,
      row,
      FOLLOWER_LOOKUP_FIELDS,
    ));
    return this.lookupWriteChain;
  }

  _isFresh(row) {
    const raw = row.last_lookup_at;
    if (!raw || this.cacheHours === 0) return false;
    const timestamp = new Date(raw).getTime();
    if (!Number.isFinite(timestamp)) return false;
    return this.now().getTime() - timestamp < this.cacheHours * 3_600_000;
  }

  _enqueue(row, force = false) {
    if (this.stopped || !row.username || this.queued.has(row._key)) return false;
    if (!force && this._isFresh(row)) return false;
    row.lookup_status = "queued";
    row.last_error = "";
    this.queue.push(row);
    this.queued.add(row._key);
    this.stats.queued += 1;
    this._pump();
    return true;
  }

  async trackUser({ userId = "", username = "", seenAt = "" }) {
    await this.ready;
    const normalizedId = String(userId ?? "").trim();
    const normalizedUsername = String(username ?? "").trim().replace(/^@/, "");
    if (!normalizedId && !normalizedUsername) return null;
    const timestamp = seenAt || this.now().toISOString();
    const row = this._findOrCreate(normalizedId, normalizedUsername, timestamp);
    this._enqueue(row, false);
    this._markUsersDirty();
    return row;
  }

  async enqueueAll({ force = false } = {}) {
    await this.ready;
    let count = 0;
    for (const row of this.users.values()) {
      if (this._enqueue(row, force)) count += 1;
    }
    this._markUsersDirty();
    return count;
  }

  _pump() {
    while (!this.stopped && this.active < this.concurrency && this.queue.length) {
      const row = this.queue.shift();
      this.active += 1;
      void this._lookupOne(row).finally(() => {
        this.active -= 1;
        this.queued.delete(row._key);
        this._pump();
        this._resolveDrainIfIdle();
      });
    }
    this._resolveDrainIfIdle();
  }

  async _lookupOne(row) {
    let result;
    try {
      result = await this.lookupImpl({ username: row.username, userId: row.user_id });
    } catch (error) {
      result = {
        status: "web_error",
        error: String(error?.message ?? error).slice(0, 500),
        source: this.source,
      };
    }
    const source = result.source || this.source;
    const collectedAt = this.now().toISOString();
    row.last_lookup_at = collectedAt;
    row.lookup_status = result.status;
    row.last_error = String(result.error ?? "").slice(0, 500);
    if (result.status === "success") {
      row.follower_count = result.followerCount;
      row.follower_count_collected_at = collectedAt;
      this.stats.success += 1;
    } else if (["not_professional_or_unavailable", "profile_unavailable"].includes(result.status)) {
      this.stats.unavailable += 1;
    } else {
      this.stats.failed += 1;
    }
    await this._scheduleLookupWrite({
      collected_at: collectedAt,
      user_id: row.user_id,
      username: row.username,
      follower_count: result.status === "success" ? result.followerCount : "",
      error: row.last_error,
    });
    this.stats.completed += 1;
    if (this.onProgress) {
      try {
        this.onProgress({
          completed: this.stats.completed,
          queued: this.stats.queued,
          username: row.username,
          status: result.status,
          followerCount: result.status === "success" ? result.followerCount : "",
          error: row.last_error,
        });
      } catch {
        // Console progress reporting must not interrupt collection.
      }
    }
    this._markUsersDirty();
    if (["rate_limited", "login_required", "challenge_required"].includes(result.status)) {
      if (!this.stats.stopStatus) {
        this.stats.stopStatus = result.status;
        this.stats.stopError = row.last_error || "Follower lookup stopped.";
      }
      this.stopped = true;
      const deferredStatus = `deferred_${result.status}`;
      for (const pending of this.queue.splice(0)) {
        pending.lookup_status = deferredStatus;
        this.queued.delete(pending._key);
      }
      this._markUsersDirty();
    }
  }

  _resolveDrainIfIdle() {
    if (this.active || this.queue.length) return;
    for (const resolve of this.drainWaiters.splice(0)) resolve();
  }

  async drain() {
    await this.ready;
    if (this.active || this.queue.length) {
      await new Promise((resolve) => this.drainWaiters.push(resolve));
    }
    await this._flushUsers();
    await this.writeChain;
    await this.lookupWriteChain;
    return { ...this.stats, stopped: this.stopped };
  }
}

export {
  FOLLOWER_LOOKUP_FIELDS,
  FollowerEnricher,
  USER_FIELDS,
  csvObjects,
  parseInstagramJson,
};

File: python_version/collectors/instagram_follower_enricher.py
"""Follower history and queue management for the Python Instagram collector."""

from __future__ import annotations

import asyncio
import csv
import os
import re
import tempfile
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Any, Awaitable, Callable


USER_FIELDS = [
    "user_id",
    "username",
    "biography",
    "follower_count",
    "collected_at",
]
FOLLOWER_SNAPSHOT_FIELDS = ["follower_count", "collected_at"]
LEGACY_FOLLOWER_TIMESTAMP_FIELD = "follower_count_collected_at"
USER_HISTORY_DIRECTORY = ".collector"
USER_HISTORY_FILENAME = "users_history_active.csv"
USER_FLUSH_CHANGE_COUNT = 500
MAX_CONSECUTIVE_WEB_ERRORS = 5
FOLLOWER_CACHE_HOURS = 6

Lookup = Callable[[dict[str, str]], Awaitable[dict[str, Any]]]


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


def isoformat_utc(value: datetime) -> str:
    return value.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def parse_timestamp(value: str) -> datetime | None:
    try:
        parsed = datetime.fromisoformat(str(value).strip().replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _exact_nonnegative_integer(value: Any) -> int | None:
    return value if isinstance(value, int) and not isinstance(value, bool) and value >= 0 else None


def _validate_exact_follower_result(result: dict[str, Any]) -> dict[str, Any]:
    validated = dict(result)
    if validated.get("status") != "success":
        return validated
    source_field = validated.get("sourceField")
    if source_field != "edge_followed_by.count":
        validated.update({
            "status": "untrusted_follower_source",
            "error": f"Rejected follower count from untrusted source field: {source_field or '<missing>'}.",
            "followerCount": None,
        })
        return validated
    exact_count = _exact_nonnegative_integer(validated.get("followerCount"))
    if exact_count is None:
        validated.update({
            "status": "invalid_follower_count",
            "error": "edge_followed_by.count was not an exact nonnegative integer.",
            "followerCount": None,
        })
        return validated
    validated["followerCount"] = exact_count
    return validated


def read_csv_objects(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    if not path.exists() or path.stat().st_size == 0:
        return [], []
    with path.open("r", newline="", encoding="utf-8-sig") as file:
        reader = csv.DictReader(file)
        return list(reader.fieldnames or []), [dict(row) for row in reader]


def atomic_write_csv(path: Path, rows: list[dict[str, Any]], fields: list[str]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    descriptor, temporary_name = tempfile.mkstemp(
        prefix=f".{path.name}.", suffix=".tmp", dir=path.parent
    )
    os.close(descriptor)
    temporary = Path(temporary_name)
    try:
        with temporary.open("w", newline="", encoding="utf-8-sig") as file:
            writer = csv.DictWriter(file, fieldnames=fields, extrasaction="ignore", lineterminator="\r\n")
            writer.writeheader()
            writer.writerows({field: row.get(field, "") for field in fields} for row in rows)
        os.replace(temporary, path)
    finally:
        temporary.unlink(missing_ok=True)


def user_history_path(data_dir: Path | str) -> Path:
    return Path(data_dir).resolve() / USER_HISTORY_DIRECTORY / USER_HISTORY_FILENAME


def ensure_user_history(data_dir: Path | str) -> Path:
    """Migrate the old public users.csv into the collector-only active history once."""
    destination = Path(data_dir).resolve()
    history_path = user_history_path(destination)
    if history_path.exists() and history_path.stat().st_size:
        return history_path
    fields, rows = read_csv_objects(destination / "users.csv")
    if fields:
        migrated_fields, migrated_rows = migrate_user_rows(fields, rows)
        atomic_write_csv(history_path, migrated_rows, migrated_fields)
    return history_path


def ordinal(number: int) -> str:
    remainder = number % 100
    if 11 <= remainder <= 13:
        return f"{number}th"
    return f"{number}{ {1: 'st', 2: 'nd', 3: 'rd'}.get(number % 10, 'th') }"


def follower_snapshot_labels(fields: list[str]) -> list[str]:
    labels: list[str] = []
    for field in fields:
        match = re.match(r"^(\d+(?:st|nd|rd|th) collect)_(.+)$", field)
        if (
            match
            and match.group(2) in {*FOLLOWER_SNAPSHOT_FIELDS, LEGACY_FOLLOWER_TIMESTAMP_FIELD}
            and match.group(1) not in labels
        ):
            labels.append(match.group(1))
    return labels


def follower_snapshot_field(label: str, field: str) -> str:
    return field if label == "Initial" else f"{label}_{field}"


def latest_follower_snapshot(row: dict[str, Any], fields: list[str]) -> dict[str, str]:
    count = str(row.get("follower_count", "") or "")
    collected_at = str(row.get("collected_at", "") or row.get(LEGACY_FOLLOWER_TIMESTAMP_FIELD, "") or "")
    for label in follower_snapshot_labels(fields):
        snapshot_count = str(row.get(follower_snapshot_field(label, "follower_count"), "") or "")
        snapshot_time = str(
            row.get(follower_snapshot_field(label, "collected_at"), "")
            or row.get(follower_snapshot_field(label, LEGACY_FOLLOWER_TIMESTAMP_FIELD), "")
            or ""
        )
        if snapshot_count:
            count = snapshot_count
        if snapshot_time:
            collected_at = snapshot_time
    return {"follower_count": count, "collected_at": collected_at}


def migrate_user_rows(
    fields: list[str], rows: list[dict[str, str]]
) -> tuple[list[str], list[dict[str, str]]]:
    """Rename the retired follower timestamp columns without losing history."""
    labels = follower_snapshot_labels(fields)
    migrated_fields = [
        *USER_FIELDS,
        *(f"{label}_{field}" for label in labels for field in FOLLOWER_SNAPSHOT_FIELDS),
    ]
    migrated_rows: list[dict[str, str]] = []
    for source in rows:
        migrated: dict[str, str] = {}
        for field in migrated_fields:
            legacy_field = (
                LEGACY_FOLLOWER_TIMESTAMP_FIELD
                if field == "collected_at"
                else field.replace("_collected_at", f"_{LEGACY_FOLLOWER_TIMESTAMP_FIELD}")
                if field.endswith("_collected_at")
                else field
            )
            migrated[field] = str(source.get(field, "") or source.get(legacy_field, "") or "")
        migrated_rows.append(migrated)
    return migrated_fields, migrated_rows


class FollowerEnricher:
    """Collect follower counts in the background and persist horizontal snapshots."""

    def __init__(
        self,
        *,
        data_dir: Path | str,
        lookup_impl: Lookup,
        concurrency: int = 1,
        source: str = "instagram_web",
        on_progress: Callable[[dict[str, Any]], Any] | None = None,
        now: Callable[[], datetime] = utc_now,
    ) -> None:
        if not callable(lookup_impl):
            raise TypeError("FollowerEnricher requires a web lookup function.")
        self.data_dir = Path(data_dir).resolve()
        self.users_path = user_history_path(self.data_dir)
        self.lookup_impl = lookup_impl
        self.concurrency = max(1, min(10, int(concurrency or 1)))
        self.source = source
        self.on_progress = on_progress
        self.now = now
        self.user_fields = list(USER_FIELDS)
        self.users: dict[str, dict[str, Any]] = {}
        self.user_id_index: dict[str, dict[str, Any]] = {}
        self.username_index: dict[str, dict[str, Any]] = {}
        self.last_lookup_at: dict[str, str] = {}
        self.queue: list[dict[str, Any]] = []
        self.queued: set[str] = set()
        self.active_tasks: set[asyncio.Task[None]] = set()
        self.stopped = False
        self.loaded = False
        self.dirty_changes = 0
        self.flush_task: asyncio.Task[None] | None = None
        self.flush_task_is_delayed = False
        self.write_lock = asyncio.Lock()
        self.consecutive_web_errors = 0
        self.stats: dict[str, Any] = {
            "queued": 0,
            "success": 0,
            "unavailable": 0,
            "failed": 0,
            "completed": 0,
            "stopStatus": "",
            "stopError": "",
        }

    async def ready(self) -> None:
        if self.loaded:
            return
        self.data_dir.mkdir(parents=True, exist_ok=True)
        ensure_user_history(self.data_dir)
        fields, rows = read_csv_objects(self.users_path)
        if fields:
            self.user_fields, rows = migrate_user_rows(fields, rows)
            if fields != self.user_fields:
                atomic_write_csv(self.users_path, rows, self.user_fields)
        for source in rows:
            row = {field: source.get(field, "") for field in self.user_fields}
            key = f"id:{row['user_id']}" if row["user_id"] else f"username:{row['username'].lower()}"
            if key.endswith(":"):
                continue
            row["_key"] = key
            self.users[key] = row
            self._index_row(row)
            latest = latest_follower_snapshot(row, self.user_fields)
            if latest["collected_at"]:
                self.last_lookup_at[key] = latest["collected_at"]
        self.loaded = True

    def set_lookup_impl(self, lookup_impl: Lookup) -> None:
        if not callable(lookup_impl):
            raise TypeError("FollowerEnricher requires a web lookup function.")
        self.lookup_impl = lookup_impl

    def _index_row(self, row: dict[str, Any], previous_username: str = "") -> None:
        if row.get("user_id"):
            self.user_id_index[str(row["user_id"])] = row
        if previous_username and self.username_index.get(previous_username.lower()) is row:
            self.username_index.pop(previous_username.lower(), None)
        if row.get("username"):
            self.username_index[str(row["username"]).lower()] = row

    def _find_or_create(self, user_id: str, username: str) -> dict[str, Any]:
        row = self.user_id_index.get(user_id) if user_id else None
        if row is None and username:
            row = self.username_index.get(username.lower())
        if row is None:
            key = f"id:{user_id}" if user_id else f"username:{username.lower()}"
            row = {field: "" for field in self.user_fields}
            row["_key"] = key
            self.users[key] = row
        previous_username = str(row.get("username", ""))
        if user_id:
            row["user_id"] = user_id
        if username:
            row["username"] = username
        self._index_row(row, previous_username)
        return row

    def _is_fresh(self, row: dict[str, Any]) -> bool:
        raw = self.last_lookup_at.get(str(row["_key"])) or latest_follower_snapshot(row, self.user_fields)["collected_at"]
        timestamp = parse_timestamp(raw) if raw else None
        return timestamp is not None and self.now().astimezone(timezone.utc) - timestamp < timedelta(hours=FOLLOWER_CACHE_HOURS)

    def _next_collection_label(self, row: dict[str, Any]) -> str:
        completed = int(bool(row.get("collected_at")))
        completed += sum(
            bool(row.get(follower_snapshot_field(label, "collected_at")))
            for label in follower_snapshot_labels(self.user_fields)
        )
        return f"{ordinal(completed + 1)} collect" if completed else "Initial"

    def _ensure_snapshot_fields(self, label: str) -> None:
        if label == "Initial":
            return
        for field in FOLLOWER_SNAPSHOT_FIELDS:
            snapshot_field = follower_snapshot_field(label, field)
            if snapshot_field not in self.user_fields:
                self.user_fields.append(snapshot_field)
                for row in self.users.values():
                    row.setdefault(snapshot_field, "")

    def _record_snapshot(self, row: dict[str, Any], label: str, count: Any, collected_at: str) -> None:
        self._ensure_snapshot_fields(label)
        row[follower_snapshot_field(label, "follower_count")] = str(count)
        row[follower_snapshot_field(label, "collected_at")] = collected_at

    def _enqueue(self, row: dict[str, Any], *, force: bool = False) -> bool:
        key = str(row["_key"])
        has_biography = bool(str(row.get("biography", "") or "").strip())
        if self.stopped or not row.get("username") or key in self.queued or (not force and has_biography and self._is_fresh(row)):
            return False
        row["lookup_status"] = "queued"
        row["last_error"] = ""
        self.queue.append(row)
        self.queued.add(key)
        self.stats["queued"] += 1
        self._pump()
        return True

    async def track_user(
        self,
        *,
        user_id: Any = "",
        username: Any = "",
        seen_at: str = "",
        enqueue: bool = True,
    ) -> dict[str, Any] | None:
        await self.ready()
        normalized_id = str(user_id or "").strip()
        normalized_username = str(username or "").strip().lstrip("@")
        if not normalized_id and not normalized_username:
            return None
        row = self._find_or_create(normalized_id, normalized_username)
        if enqueue:
            self._enqueue(row)
        self._mark_dirty()
        latest = latest_follower_snapshot(row, self.user_fields)
        return {**row, **latest}

    async def enqueue_all(self) -> int:
        await self.ready()
        count = sum(1 for row in list(self.users.values()) if self._enqueue(row))
        self._mark_dirty()
        return count

    async def enqueue_all_exact(self) -> int:
        """Refresh every known user even when a legacy snapshot is still fresh."""
        await self.ready()
        count = sum(1 for row in list(self.users.values()) if self._enqueue(row, force=True))
        self._mark_dirty()
        return count

    async def lookup_user_now(
        self,
        *,
        user_id: Any = "",
        username: Any = "",
        seen_at: str = "",
    ) -> dict[str, Any]:
        """Run a fresh lookup immediately, bypassing legacy cached values."""
        del seen_at
        await self.ready()
        normalized_id = str(user_id or "").strip()
        normalized_username = str(username or "").strip().lstrip("@")
        if not normalized_id and not normalized_username:
            return {"status": "profile_unavailable", "error": "Missing Instagram user identity.", "follower_count": ""}
        row = self._find_or_create(normalized_id, normalized_username)
        self.stats["queued"] += 1
        result = await self._lookup_one(row)
        latest = latest_follower_snapshot(row, self.user_fields)
        succeeded = result.get("status") == "success"
        return {
            **row,
            **latest,
            "status": str(result.get("status", "web_error")),
            "error": str(result.get("error", "")),
            "followerCount": result.get("followerCount") if succeeded else None,
            "sourceField": result.get("sourceField"),
        }

    def _pump(self) -> None:
        while not self.stopped and self.queue and len(self.active_tasks) < self.concurrency:
            row = self.queue.pop(0)
            task = asyncio.create_task(self._lookup_one(row))
            self.active_tasks.add(task)
            task.add_done_callback(self._task_finished)

    def _task_finished(self, task: asyncio.Task[None]) -> None:
        self.active_tasks.discard(task)
        try:
            task.result()
        except Exception:
            # _lookup_one converts lookup failures to a result; this only protects the queue.
            pass
        self._pump()

    async def _lookup_one(self, row: dict[str, Any]) -> dict[str, Any]:
        key = str(row["_key"])
        label = self._next_collection_label(row)
        try:
            result = await self.lookup_impl({"username": str(row.get("username", "")), "userId": str(row.get("user_id", ""))})
        except Exception as error:
            result = {"status": "web_error", "error": str(error)[:500], "source": self.source}
        result = _validate_exact_follower_result(result)
        collected_at = isoformat_utc(self.now())
        self.last_lookup_at[key] = collected_at
        status = str(result.get("status", "web_error"))
        row["lookup_status"] = status
        row["last_error"] = str(result.get("error", ""))[:500]
        if "biography" in result:
            row["biography"] = str(result.get("biography") or "")
        if status == "success":
            self._record_snapshot(row, label, result.get("followerCount", ""), collected_at)
            self.stats["success"] += 1
        elif status in {"not_professional_or_unavailable", "profile_unavailable"}:
            self.stats["unavailable"] += 1
        else:
            self.stats["failed"] += 1
        self.consecutive_web_errors = self.consecutive_web_errors + 1 if status == "web_error" else 0
        self.stats["completed"] += 1
        if self.on_progress:
            try:
                self.on_progress({
                    "completed": self.stats["completed"],
                    "queued": self.stats["queued"],
                    "username": row.get("username", ""),
                    "status": status,
                    "followerCount": result.get("followerCount", "") if status == "success" else "",
                    "error": row["last_error"],
                })
            except Exception:
                pass
        self._mark_dirty()
        repeated_error = self.consecutive_web_errors >= MAX_CONSECUTIVE_WEB_ERRORS
        if status in {"rate_limited", "login_required", "challenge_required"} or repeated_error:
            if not self.stats["stopStatus"]:
                self.stats["stopStatus"] = "repeated_web_error" if repeated_error else status
                self.stats["stopError"] = (
                    f"{MAX_CONSECUTIVE_WEB_ERRORS} consecutive follower web errors."
                    if repeated_error else row["last_error"] or "Follower lookup stopped."
                )
            self.stopped = True
            deferred_status = f"deferred_{self.stats['stopStatus']}"
            for pending in self.queue:
                pending["lookup_status"] = deferred_status
                self.queued.discard(str(pending["_key"]))
            self.queue.clear()
            self._mark_dirty()
        self.queued.discard(key)
        return result

    def _mark_dirty(self) -> None:
        self.dirty_changes += 1
        if self.dirty_changes >= USER_FLUSH_CHANGE_COUNT:
            if self.flush_task and not self.flush_task.done() and self.flush_task_is_delayed:
                self.flush_task.cancel()
                self.flush_task = None
            if self.flush_task is None or self.flush_task.done():
                self.flush_task_is_delayed = False
                self.flush_task = asyncio.create_task(self._flush_users())
        elif self.flush_task is None or self.flush_task.done():
            self.flush_task_is_delayed = True
            self.flush_task = asyncio.create_task(self._delayed_flush())

    async def _delayed_flush(self) -> None:
        await asyncio.sleep(60)
        self.flush_task_is_delayed = False
        await self._flush_users()

    async def _flush_users(self) -> None:
        if not self.dirty_changes:
            return
        async with self.write_lock:
            while self.dirty_changes:
                self.dirty_changes = 0
                fields = list(self.user_fields)
                rows = [
                    {field: row.get(field, "") for field in fields}
                    for row in sorted(self.users.values(), key=lambda item: str(item.get("username", "")).lower())
                ]
                await asyncio.to_thread(atomic_write_csv, self.users_path, rows, fields)

    async def drain(self) -> dict[str, Any]:
        await self.ready()
        while self.queue or self.active_tasks:
            self._pump()
            if self.active_tasks:
                await asyncio.wait(self.active_tasks, return_when=asyncio.FIRST_COMPLETED)
            else:
                await asyncio.sleep(0)
        if self.flush_task and not self.flush_task.done():
            if self.flush_task_is_delayed:
                self.flush_task.cancel()
                try:
                    await self.flush_task
                except asyncio.CancelledError:
                    pass
            else:
                await self.flush_task
        await self._flush_users()
        return {**self.stats, "stopped": self.stopped}
